File: wrappers/java/src/main/java/com/evernym/vdrtools/vdr/VDR.java
package com.evernym.vdrtools.vdr;

import com.evernym.vdrtools.IndyException;
import com.evernym.vdrtools.IndyJava;
import com.evernym.vdrtools.LibIndy;
import com.evernym.vdrtools.ParamGuard;
import com.sun.jna.Callback;
import com.sun.jna.Pointer;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;


public class VDR extends IndyJava.API implements AutoCloseable {

    private final int vdrHandle;

    private VDR(int vdrHandle) {
        this.vdrHandle = vdrHandle;
    }

    public int getVdrHandle() {
        return vdrHandle;
    }

    private static Callback stringCb = new Callback() {

        @SuppressWarnings({"unused", "unchecked"})
        public void callback(int xcommand_handle, int err, String str) {
            CompletableFuture<String> future = (CompletableFuture<String>) removeFuture(xcommand_handle);
            if (!checkResult(future, err)) return;

            String result = str;
            future.complete(result);
        }
    };

    private static Callback voidCb = new Callback() {

        @SuppressWarnings({"unused", "unchecked"})
        public void callback(int xcommand_handle, int err) {

            CompletableFuture<Void> future = (CompletableFuture<Void>) removeFuture(xcommand_handle);
            if (!checkResult(future, err)) return;

            Void result = null;
            future.complete(result);
        }
    };

    private static Callback createVdrCb = new Callback() {

        @SuppressWarnings({"unused", "unchecked"})
        public void callback(int xcommand_handle, int err, int vdr_handle) {

            CompletableFuture<VDR> future = (CompletableFuture<VDR>) removeFuture(xcommand_handle);
            if (!checkResult(future, err)) return;

            VDR result = new VDR(vdr_handle);
            future.complete(result);
        }
    };

    private static Callback prepareTxnCb = new Callback() {

        @SuppressWarnings({"unused", "unchecked"})
        public void callback(int xcommand_handle, int err, String namespace, String signature_spec,Pointer txn_bytes_raw, int txn_bytes_len, Pointer bytes_to_sign_raw, int bytes_to_sign_len, String endorsement_spec) {

            CompletableFuture<VdrResults.PreparedTxnResult> future = (CompletableFuture<VdrResults.PreparedTxnResult>) removeFuture(xcommand_handle);
            if (!checkResult(future, err)) return;

            byte[] txnBytes = new byte[bytes_to_sign_len];
            bytes_to_sign_raw.read(0, txnBytes, 0, bytes_to_sign_len);

            byte[] bytesToSign = new byte[bytes_to_sign_len];
            bytes_to_sign_raw.read(0, bytesToSign, 0, bytes_to_sign_len);

            VdrResults.PreparedTxnResult result = new VdrResults.PreparedTxnResult(namespace, signature_spec,txnBytes, bytesToSign, endorsement_spec);
            future.complete(result);
        }
    };

    public static CompletableFuture<VDR> createVDR() throws IndyException {
        CompletableFuture<VDR> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int result = LibIndy.api.vdr_create(
                commandHandle,
                createVdrCb
        );

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<Void> registerIndyLedger(
            VDR vdr,
            String namespaceList,
            String genesisTxnData,
            String taaConfig
    ) throws IndyException {
        ParamGuard.notNull(namespaceList, "namespaceList");
        ParamGuard.notNull(genesisTxnData, "genesisTxnData");
        ParamGuard.notNull(taaConfig, "taaConfig");

        CompletableFuture<Void> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_register_indy_ledger(
                commandHandle,
                handle,
                namespaceList,
                genesisTxnData,
                taaConfig,
                voidCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<Void> registerCheqdLedger(
            VDR vdr,
            String namespaceList,
            String chainId,
            String nodeAddrList
    ) throws IndyException {
        ParamGuard.notNull(namespaceList, "namespaceList");
        ParamGuard.notNull(chainId, "chainId");
        ParamGuard.notNull(nodeAddrList, "nodeAddrList");

        CompletableFuture<Void> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_register_cheqd_ledger(
                commandHandle,
                handle,
                namespaceList,
                chainId,
                nodeAddrList,
                voidCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<String> ping(
            VDR vdr,
            String namespaceList
    ) throws IndyException {
        ParamGuard.notNull(namespaceList, "namespaceList");

        CompletableFuture<String> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_ping(
                commandHandle,
                handle,
                namespaceList,
                stringCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<Void> cleanup(
            VDR vdr
    ) throws IndyException {
        CompletableFuture<Void> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();
        int result = LibIndy.api.vdr_cleanup(
                commandHandle,
                handle,
                voidCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<String> resolveDID(
            VDR vdr,
            String fqDID,
            String cacheOptions
    ) throws IndyException {
        ParamGuard.notNull(fqDID, "fqDID");
        ParamGuard.notNull(cacheOptions, "cacheOptions");

        CompletableFuture<String> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_resolve_did(
                commandHandle,
                handle,
                fqDID,
                cacheOptions,
                stringCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<String> resolveSchema(
            VDR vdr,
            String fqSchema,
            String cacheOptions
    ) throws IndyException {
        ParamGuard.notNull(fqSchema, "fqSchema");
        ParamGuard.notNull(cacheOptions, "cacheOptions");

        CompletableFuture<String> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_resolve_schema(
                commandHandle,
                handle,
                fqSchema,
                cacheOptions,
                stringCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<String> resloveCredDef(
            VDR vdr,
            String fqCredDef,
            String cacheOptions
    ) throws IndyException {
        ParamGuard.notNull(fqCredDef, "fqCredDef");
        ParamGuard.notNull(cacheOptions, "cacheOptions");

        CompletableFuture<String> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_resolve_cred_def(
                commandHandle,
                handle,
                fqCredDef,
                cacheOptions,
                stringCb);

        checkResult(future, result);

        return future;
    }


    private static CompletableFuture<VdrResults.PreparedTxnResult> prepareDID(
            VDR vdr,
            String txnSpecificParams,
            String submitterDID,
            String endorser
    ) throws IndyException {
        ParamGuard.notNull(txnSpecificParams, "txnSpecificParams");
        ParamGuard.notNull(submitterDID, "submitterDID");

        CompletableFuture<VdrResults.PreparedTxnResult> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_prepare_did(
                commandHandle,
                handle,
                txnSpecificParams,
                submitterDID,
                endorser,
                stringCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<VdrResults.PreparedTxnResult> prepareSchema(
            VDR vdr,
            String txnSpecificParams,
            String submitterDID,
            String endorser
    ) throws IndyException {
        ParamGuard.notNull(txnSpecificParams, "txnSpecificParams");
        ParamGuard.notNull(submitterDID, "submitterDID");

        CompletableFuture<VdrResults.PreparedTxnResult> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_prepare_schema(
                commandHandle,
                handle,
                txnSpecificParams,
                submitterDID,
                endorser,
                stringCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<VdrResults.PreparedTxnResult> prepareCredDef(
            VDR vdr,
            String txnSpecificParams,
            String submitterDID,
            String endorser
    ) throws IndyException {
        ParamGuard.notNull(txnSpecificParams, "txnSpecificParams");
        ParamGuard.notNull(submitterDID, "submitterDID");

        CompletableFuture<VdrResults.PreparedTxnResult> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_prepare_cred_def(
                commandHandle,
                handle,
                txnSpecificParams,
                submitterDID,
                endorser,
                stringCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<String> submitTxn(
            VDR vdr,
            VdrResults.PreparedTxnResult preparedTxn,
            byte[] signature
    ) throws IndyException {
        ParamGuard.notNull(preparedTxn, "preparedTxn");
        ParamGuard.notNull(signature, "signature");

        CompletableFuture<String> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_submit_txn(
                commandHandle,
                handle,
                preparedTxn.getNamespace(),
                preparedTxn.getSignatureSpec(),
                preparedTxn.getTxnBytes(),
                preparedTxn.getTxnBytes().length,
                signature,
                signature.length,
                preparedTxn.getEndorsementSpec(),
                stringCb);

        checkResult(future, result);

        return future;
    }

    private static CompletableFuture<String> submitQuery(
            VDR vdr,
            String namespace,
            String query
    ) throws IndyException {
        ParamGuard.notNull(namespace, "namespace");
        ParamGuard.notNull(query, "query");

        CompletableFuture<String> future = new CompletableFuture<>();
        int commandHandle = addFuture(future);

        int handle = vdr.getVdrHandle();

        int result = LibIndy.api.vdr_submit_query(
                commandHandle,
                handle,
                namespace,
                query,
                stringCb
        );

        checkResult(future,result);

        return future;
    }


    public CompletableFuture<Void> registerIndyLedger(
            String namespaceList,
            String genesisTxnData,
            String taaConfig
    ) throws IndyException {
        return registerIndyLedger(this, namespaceList, genesisTxnData, taaConfig);
    }


    public CompletableFuture<Void> registerCheqdLedger(
            String namespaceList,
            String chainId,
            String nodeAddrList
    ) throws IndyException {
        return registerCheqdLedger(this, namespaceList, chainId, nodeAddrList);
    }

    public CompletableFuture<String> ping(
            String namespaceList
    ) throws IndyException {
        return ping(this, namespaceList);
    }

    public CompletableFuture<Void> cleanup() throws IndyException {
        return cleanup(this);
    }

    public CompletableFuture<String> resolveDID(
            String fqDID,
            String cacheOptions
    ) throws IndyException {
        return resolveDID(this, fqDID, cacheOptions);
    }

    public CompletableFuture<String> resolveSchema(
            String fqSchema,
            String cacheOptions
    ) throws IndyException {
        return resolveSchema(this, fqSchema, cacheOptions);
    }

    public CompletableFuture<String> resloveCredDef(
            String fqCredDef,
            String cacheOptions
    ) throws IndyException {
        return resloveCredDef(this, fqCredDef, cacheOptions);
    }

    public CompletableFuture<VdrResults.PreparedTxnResult> prepareDID(
            String txnSpecificParams,
            String submitterDID,
            String endorser
    ) throws IndyException {
        return prepareDID(this, txnSpecificParams, submitterDID, endorser);
    }

    public CompletableFuture<VdrResults.PreparedTxnResult> prepareSchema(
            String txnSpecificParams,
            String submitterDID,
            String endorser
    ) throws IndyException {
        return prepareSchema(this, txnSpecificParams, submitterDID, endorser);
    }

    public CompletableFuture<VdrResults.PreparedTxnResult> prepareCredDef(
            String txnSpecificParams,
            String submitterDID,
            String endorser
    ) throws IndyException {
        return prepareCredDef(this, txnSpecificParams, submitterDID, endorser);
    }

    public CompletableFuture<String> submitTxn(
            VdrResults.PreparedTxnResult preparedTxn,
            byte[] signature
    ) throws IndyException {
        return submitTxn(this, preparedTxn, signature);
    }

    public CompletableFuture<String> submitQuery(
            String namespace,
            String query
    ) throws IndyException {
        return submitQuery(this, namespace, query);
    }

    @Override
    public void close() throws InterruptedException, ExecutionException, IndyException {
        cleanup().get();
    }
}

File: wrappers/rust/src/vdr.rs
use std::ffi::CString;

use futures::Future;

use crate::{ErrorCode, IndyError};
use crate::{CommandHandle, VdrHandle};
use crate::ffi::{ResponseEmptyCB, ResponsePreparedTxnCB, ResponseStringCB, ResponseVdrHandleCB};
use crate::ffi::vdr;
use crate::utils::callbacks::{ClosureHandler, ResultHandler};

pub fn vdr_create() -> Box<dyn Future<Item=VdrHandle, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_vdrhandle();

    let err = _vdr_create(command_handle, cb);

    ResultHandler::vdrhandle(command_handle, err, receiver)
}

fn _vdr_create(command_handle: CommandHandle, cb: Option<ResponseVdrHandleCB>) -> ErrorCode {
    ErrorCode::from(unsafe {
        vdr::vdr_create(command_handle, cb)
    })
}

pub fn vdr_register_indy_ledger(vdr_handle: VdrHandle, namespace_list: &str, genesis_txn_data: &str, taa_config: &str) -> Box<dyn Future<Item=(), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec();

    let err = _vdr_register_indy_ledger(command_handle, vdr_handle, namespace_list, genesis_txn_data, taa_config, cb);

    ResultHandler::empty(command_handle, err, receiver)
}

fn _vdr_register_indy_ledger(command_handle: CommandHandle, vdr_handle: VdrHandle, namespace_list: &str, genesis_txn_data: &str, taa_config: &str, cb: Option<ResponseEmptyCB>) -> ErrorCode {
    let namespace_list = c_str!(namespace_list);
    let genesis_txn_data = c_str!(genesis_txn_data);
    let taa_config = c_str!(taa_config);
    ErrorCode::from(unsafe {
        vdr::vdr_register_indy_ledger(command_handle, vdr_handle, namespace_list.as_ptr(), genesis_txn_data.as_ptr(), taa_config.as_ptr(), cb)
    })
}

pub fn vdr_register_cheqd_ledger(vdr_handle: VdrHandle, namespace_list: &str, chain_id: &str, node_addrs_list: &str) -> Box<dyn Future<Item=(), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec();

    let err = _vdr_register_cheqd_ledger(command_handle, vdr_handle, namespace_list, chain_id, node_addrs_list, cb);

    ResultHandler::empty(command_handle, err, receiver)
}

fn _vdr_register_cheqd_ledger(command_handle: CommandHandle, vdr_handle: VdrHandle, namespace_list: &str, chain_id: &str, node_addrs_list: &str, cb: Option<ResponseEmptyCB>) -> ErrorCode {
    let namespace_list = c_str!(namespace_list);
    let chain_id = c_str!(chain_id);
    let node_addrs_list = c_str!(node_addrs_list);

    ErrorCode::from(unsafe {
        vdr::vdr_register_indy_ledger(command_handle, vdr_handle, namespace_list.as_ptr(), chain_id.as_ptr(), node_addrs_list.as_ptr(), cb)
    })
}

pub fn vdr_ping(vdr_handle: VdrHandle, namespace_list: &str) -> Box<dyn Future<Item=String, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_string();

    let err = _vdr_ping(command_handle, vdr_handle, namespace_list, cb);

    ResultHandler::str(command_handle, err, receiver)
}

fn _vdr_ping(command_handle: CommandHandle, vdr_handle: VdrHandle, namespace_list: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
    let namespace_list = c_str!(namespace_list);

    ErrorCode::from(unsafe {
        vdr::vdr_ping(command_handle, vdr_handle, namespace_list.as_ptr(), cb)
    })
}

pub fn vdr_cleanup(vdr_handle: VdrHandle) -> Box<dyn Future<Item=(), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec();

    let err = _vdr_cleanup(command_handle, vdr_handle, cb);

    ResultHandler::empty(command_handle, err, receiver)
}

fn _vdr_cleanup(command_handle: CommandHandle, vdr_handle: VdrHandle, cb: Option<ResponseEmptyCB>) -> ErrorCode {
    ErrorCode::from(unsafe {
        vdr::vdr_cleanup(command_handle, vdr_handle, cb)
    })
}

pub fn vdr_resolve_did(vdr_handle: VdrHandle, fqdid: &str, cache_options: &str) -> Box<dyn Future<Item=String, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_string();

    let err = _vdr_resolve_did(command_handle, vdr_handle, fqdid, cache_options, cb);

    ResultHandler::str(command_handle, err, receiver)
}

fn _vdr_resolve_did(command_handle: CommandHandle, vdr_handle: VdrHandle, fqdid: &str, cache_options: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
    let fqdid = c_str!(fqdid);
    let cache_options = c_str!(cache_options);

    ErrorCode::from(unsafe {
        vdr::vdr_resolve_did(command_handle, vdr_handle, fqdid.as_ptr(), cache_options.as_ptr(), cb)
    })
}

pub fn vdr_resolve_schema(vdr_handle: VdrHandle, fqschema: &str, cache_options: &str) -> Box<dyn Future<Item=String, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_string();

    let err = _vdr_resolve_schema(command_handle, vdr_handle, fqschema, cache_options, cb);

    ResultHandler::str(command_handle, err, receiver)
}

fn _vdr_resolve_schema(command_handle: CommandHandle, vdr_handle: VdrHandle, fqschema: &str, cache_options: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
    let fqschema = c_str!(fqschema);
    let cache_options = c_str!(cache_options);

    ErrorCode::from(unsafe {
        vdr::vdr_resolve_schema(command_handle, vdr_handle, fqschema.as_ptr(), cache_options.as_ptr(), cb)
    })
}

pub fn vdr_resolve_cred_def(vdr_handle: VdrHandle, fqcreddef: &str, cache_options: &str) -> Box<dyn Future<Item=String, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_string();

    let err = _vdr_resolve_cred_def(command_handle, vdr_handle, fqcreddef, cache_options, cb);

    ResultHandler::str(command_handle, err, receiver)
}

fn _vdr_resolve_cred_def(command_handle: CommandHandle, vdr_handle: VdrHandle, fqcreddef: &str, cache_options: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
    let fqcreddef = c_str!(fqcreddef);
    let cache_options = c_str!(cache_options);

    ErrorCode::from(unsafe {
        vdr::vdr_resolve_cred_def(command_handle, vdr_handle, fqcreddef.as_ptr(), cache_options.as_ptr(), cb)
    })
}

pub fn vdr_prepare_did(vdr_handle: VdrHandle, txn_specific_params: &str, submitter_did: &str, endorser: &str) -> Box<dyn Future<Item=(String, String, Vec<u8>, Vec<u8>, String), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_preparedtxnhandle();

    let err = _vdr_prepare_did(command_handle, vdr_handle, txn_specific_params, submitter_did, endorser, cb);

    ResultHandler::preparedtxnhandle(command_handle, err, receiver)
}

fn _vdr_prepare_did(command_handle: CommandHandle, vdr_handle: VdrHandle, txn_specific_params: &str, submitter_did: &str, endorser: &str, cb: Option<ResponsePreparedTxnCB>) -> ErrorCode {
    let txn_specific_params = c_str!(txn_specific_params);
    let submitter_did = c_str!(submitter_did);
    let endorser = c_str!(endorser);

    ErrorCode::from(unsafe {
        vdr::vdr_prepare_did(command_handle, vdr_handle, txn_specific_params.as_ptr(), submitter_did.as_ptr(), endorser.as_ptr(), cb)
    })
}

pub fn vdr_prepare_schema(vdr_handle: VdrHandle, txn_specific_params: &str, submitter_schema: &str, endorser: &str) -> Box<dyn Future<Item=(String, String, Vec<u8>, Vec<u8>, String), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_preparedtxnhandle();

    let err = _vdr_prepare_schema(command_handle, vdr_handle, txn_specific_params, submitter_schema, endorser, cb);

    ResultHandler::preparedtxnhandle(command_handle, err, receiver)
}

fn _vdr_prepare_schema(command_handle: CommandHandle, vdr_handle: VdrHandle, txn_specific_params: &str, submitter_schema: &str, endorser: &str, cb: Option<ResponsePreparedTxnCB>) -> ErrorCode {
    let txn_specific_params = c_str!(txn_specific_params);
    let submitter_schema = c_str!(submitter_schema);
    let endorser = c_str!(endorser);

    ErrorCode::from(unsafe {
        vdr::vdr_prepare_schema(command_handle, vdr_handle, txn_specific_params.as_ptr(), submitter_schema.as_ptr(), endorser.as_ptr(), cb)
    })
}

pub fn vdr_prepare_cred_def(vdr_handle: VdrHandle, txn_specific_params: &str, submitter_cred_def: &str, endorser: &str) -> Box<dyn Future<Item=(String, String, Vec<u8>, Vec<u8>, String), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_preparedtxnhandle();

    let err = _vdr_prepare_cred_def(command_handle, vdr_handle, txn_specific_params, submitter_cred_def, endorser, cb);

    ResultHandler::preparedtxnhandle(command_handle, err, receiver)
}

fn _vdr_prepare_cred_def(command_handle: CommandHandle, vdr_handle: VdrHandle, txn_specific_params: &str, submitter_cred_def: &str, endorser: &str, cb: Option<ResponsePreparedTxnCB>) -> ErrorCode {
    let txn_specific_params = c_str!(txn_specific_params);
    let submitter_cred_def = c_str!(submitter_cred_def);
    let endorser = c_str!(endorser);

    ErrorCode::from(unsafe {
        vdr::vdr_prepare_cred_def(command_handle, vdr_handle, txn_specific_params.as_ptr(), submitter_cred_def.as_ptr(), endorser.as_ptr(), cb)
    })
}

pub fn vdr_submit_txn(
    vdr_hanlde: VdrHandle,
    namespace: &str,
    signature_spec: &str,
    txn_bytes: &[u8],
    signature: &[u8],
    endorsement_spec: &str,
) -> Box<dyn Future<Item=String, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_string();

    let err = _vdr_submit_txn(command_handle, vdr_hanlde, namespace, signature_spec, txn_bytes, signature, endorsement_spec, cb);

    ResultHandler::str(command_handle, err, receiver)
}

fn _vdr_submit_txn(
    command_handle: CommandHandle,
    vdr_hanlde: VdrHandle,
    namespace: &str,
    signature_spec: &str,
    txn_bytes: &[u8],
    signature: &[u8],
    endorsement_spec: &str,
    cb: Option<ResponseStringCB>,
) -> ErrorCode {
    let namespace = c_str!(namespace);
    let signature_spec = c_str!(signature_spec);
    let endorsement_spec = c_str!(endorsement_spec);

    ErrorCode::from(unsafe {
        vdr::vdr_submit_txn(command_handle, vdr_hanlde, namespace.as_ptr(), signature_spec.as_ptr(),
                            txn_bytes.as_ptr() as *const u8, txn_bytes.len() as u32,
                            signature.as_ptr() as *const u8, signature.len() as u32,
                            endorsement_spec.as_ptr(), cb)
    })
}
